waspc/src/Analyzer/Parser/Util.hs

```diff
@@ -0,0 +1,53 @@
+module Analyzer.Parser.Util
+  ( ParseState (..),
+    initialState,
+    Parser,
+    updatePosn,
+    putInput,
+    AlexInput,
+  )
+where
+
+import Analyzer.Parser.Syntax (ParseError, Posn (..))
+import Control.Monad.Trans.Except (Except)
+import Control.Monad.Trans.State.Lazy (StateT, get, modify)
+import Data.Word (Word8)
+
+-- | Tracks state of the parser, which is
+--   - Current line/column position
+--   - Current input to the lexer
+data ParseState = ParseState
+  { psPosn :: Posn,
+    psInput :: AlexInput
+  }
+  deriving (Show)
+
+-- | Create an initial state from a source string
+initialState :: String -> ParseState
+initialState source =
+  ParseState
+    { psPosn = Posn 1 1,
+      psInput = ('\n', [], source)
+    }
+
+type Parser a = StateT ParseState (Except ParseError) a
```

Comment:
type vs newtype -> makes sense, although you could just derive the needed stuff for the newtype. This is how I was doing it in hlox: https://github.com/Martinsos/lox-haskell/blob/master/src/Interpreter/Core.hs#L28 . I see that in hlox I used type on some places, newtype on some other, and I can't remember the arguments I had. I believe I was first using type until I learned I could use newtype. And I guess I liked newtype better because well, it provides some extra abstraction hm. That said, I am still not arguing it has to be newtype here, just wanted to share what I know about the topic. Personally I feel like I don't know what is better yet. `type` is simpler if nothing else.

Order of State and Except: I am asking because I know I deliberately put them in specific order here: https://github.com/Martinsos/lox-haskell/blob/master/src/Interpreter/Core.hs#L28 , after some research. I guess it is what you said -> it still gives you access to the state. But what kind of state hm -> state at the point when error happened I guess? Ok I found this SO -> https://stackoverflow.com/a/31839103/1509394 -> so this is relevant I think, when it talks in the comment about state being "rolled-back" on catchError vs not being rolled-back. That makes sense -> the question is, when we catch error, do we want the state of the parser to be as it was before the error, or as it was when error was thrown? Hmm. So right now we have the first behaviour -> that sounds like a good choice then? I wonder why I choose the second one. What do you think about this?

Reply:
State "rolling back" with `catchError` is interesting, but without any error catching (yet) it's hard to say what's desired. But right now, the order causes no difference in return type (because of `evalStateT`) or behavior, so I think not changing it is fine.